apps/api/src/scraper/scrapeURL/index.ts

```diff
@@ -0,0 +1,263 @@
+import { Logger } from "winston";
+import * as Sentry from "@sentry/node";
+
+import { Document, ScrapeOptions } from "../../controllers/v1/types";
+import { logger, ArrayTransport } from "../../lib/logger";
+import { buildFallbackList, Engine, EngineScrapeResult, FeatureFlag, scrapeURLWithEngine } from "./engines";
+import { parseMarkdown } from "../../lib/html-to-markdown";
+import { AddFeatureError, EngineError, NoEnginesLeftError, TimeoutError } from "./error";
+import { executeTransformers } from "./transformers";
+import { LLMRefusalError } from "./transformers/llmExtract";
+import { urlSpecificParams } from "./lib/urlSpecificParams";
+
+export type ScrapeUrlResponse = ({
+    success: true,
+    document: Document,
+} | {
+    success: false,
+    error: any,
+}) & {
+    logs: any[],
+}
+
+export type Meta = {
+    id: string;
+    url: string;
+    options: ScrapeOptions;
+    internalOptions: InternalOptions;
+    logger: Logger;
+    logs: any[];
+    featureFlags: Set<FeatureFlag>;
+}
+
+function buildFeatureFlags(url: string, options: ScrapeOptions, internalOptions: InternalOptions): Set<FeatureFlag> {
+    const flags: Set<FeatureFlag> = new Set();
+
+    if (options.actions !== undefined) {
+        flags.add("actions");
+    }
+
+    if (options.formats.includes("screenshot")) {
+        flags.add("screenshot");
+    }
+
+    if (options.formats.includes("screenshot@fullPage")) {
+        flags.add("screenshot@fullScreen");
+    }
+
+    if (options.waitFor !== 0) {
+        flags.add("waitFor");
+    }
+
+    if (internalOptions.atsv) {
+        flags.add("atsv");
+    }
+
+    if (options.location || options.geolocation) {
+        flags.add("location");
+    }
+
+    if (options.mobile) {
+        flags.add("mobile");
+    }
+
+    const urlO = new URL(url);
+
+    if (urlO.pathname.endsWith(".pdf")) {
+        flags.add("pdf");
+    }
+
+    if (urlO.pathname.endsWith(".docx")) {
+        flags.add("docx");
+    }
+
+    return flags;
+}
+
+function buildMetaObject(id: string, url: string, options: ScrapeOptions, internalOptions: InternalOptions): Meta {
```

Comment:
@mogery can you add a comment explaining the idea of a meta object + why we remove www?

Reply:
We do not remove `www`, only for matching the url specific options object, which is the same format as in WebScraper. Adding comment.

Reply:
can you explain the meta concept?

Reply:
Adding comment.

Reply:
Added